Adding support for ComputeCpp compiler (#13)
codeplaysoftware/cuda-to-sycl-nbody: README.md

# nbody
Accelerated N-body sim with OpenGL graphics & automatic CUDA->SYCL conversion using [dpct](https://www.intel.com/content/www/us/en/developer/tools/oneapi/dpc-compatibility-tool.html).

![](http://i.imgur.com/drzi33P.jpg)

Forked from https://github.com/salel/nbody

## Compilers/Backends

This nbody simulation can be run with any of:
 - CUDA
 - DPC++ CUDA backend
 - DPC++ OpenCL CPU backend

>  - ComputeCpp Experimental Host Backend

Reviewer: ```suggestion
 - ComputeCpp OpenCL Backend
```

sudo apt update
sudo apt install libglew-dev libglfw3-dev libglm-dev libxxf86vm-dev libxcursor-dev libxinerama-dev libxi-dev
```

The implementation relies on OpenGL 4.5.

### Simulation Dependencies (CUDA & SYCL)

The CUDA version of this code requires the [CUDA runtime](https://intel.github.io/llvm-docs/GetStartedGuide.html#build-dpc-toolchain-with-support-for-nvidia-cuda) to be installed on your machine.

The DPC++ CUDA backend version also requires the CUDA runtime.

The DPC++ OpenCL backend requires an [OpenCL runtime](https://intel.github.io/llvm-docs/GetStartedGuide.html#install-low-level-runtime). To run specifically on the CPU, you must install the OpenCL runtime for your CPU.

Both DPC++ backends require the [DPC++ compiler](https://intel.github.io/llvm-docs/GetStartedGuide.html) to compile the SYCL code.

> The ComputeCpp Experimental Host backend requires the proprietary ComputeCpp binary which you should have received from Codeplay, and an OpenCL runtime.

Reviewer: ```suggestion
The ComputeCpp backend requires ComputeCpp which you can download from the [Codeplay website](https://developer.codeplay.com/home/).
```

```
    SYCL_DEVICE_FILTER=opencl:cpu ./nbodygl
```
will run on a CPU through the OpenCL backend. Note the correspondence between options for `SYCL_DEVICE_FILTER` and the output of `sycl-ls`.

**Note**: Selection between DPC++ backends at runtime is possible because `CMakeLists.txt` specifies building the SYCL code for both CUDA (`nvptx64-nvidia-cuda`) & OpenCL (`spir64`) targets:
```
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsycl -fsycl-targets=spir64,nvptx64-nvidia-cuda -fsycl-unnamed-lambda")
```

### Selecting the ComputeCpp Host Backend

> The experimental ComputeCpp distribution will *only* work with the host backend, and this must be specified with the `COMPUTECPP_TARGET` environment variable:

Reviewer: Not 100% sure what to say here but if the user downloads ComputeCpp, it should work with either the host device or any USM-supporting OpenCL device (e.g. Intel GPU/CPU).
